generator/feature_segments_checker/feature_segments_checker.cpp

```diff
@@ -0,0 +1,274 @@
+  set<RoughPoint> m_uniqueRoadPoints;
+  /// Key is altitude difference for feature in meters. If a feature goes up the key is more then 0.
+  /// Value is a feature counter.
+  map<generator::SrtmTile::THeight, uint32_t> m_altitudeDiffs;
+  /// Key is length of feature in meters. Value is a feature counter.
+  map<uint32_t, uint32_t> m_featureLength;
+  /// Key is length of feature segment in meters. Value is a segment counter.
+  map<uint32_t, uint32_t> m_segLength;
+  /// Key is difference between two values:
+  /// 1. how many meters it's necessary to go up following the feature. If feature wavy
+  ///   calculates only raise meters.
+  /// 2. the difference in altitude between end and start of features.
+  /// Value is a segment counter.
+  map<int32_t, uint32_t> m_featureWave;
+  /// Key is number of meters which is necessary to go up following the feature.
+  /// Value is a feature counter.
+  map<int32_t, uint32_t> m_featureUp;
+  /// Key is number of meters which is necessary to go down following the feature.
+  /// Value is a feature counter.
+  map<int32_t, uint32_t> m_featureDown;
+  /// Key is number of meters. It shows altitude deviation of intermediate feature points
+  /// from linear model.
+  /// Value is a feature counter.
+  map<int32_t, uint32_t> m_diffFromLinear;
+  /// Feature counter for GetBicycleModel().IsRoad(feature) == true.
+  uint32_t m_roadCount;
+  /// Feature counter for empty features with GetBicycleModel().IsRoad(feature).
+  uint32_t m_emptyRoadCount;
+  /// Point counter for feature where GetBicycleModel().IsRoad(feature) == true.
+  uint32_t m_roadPointCount;
+  /// Feature counter for GetBicycleModel().IsRoad(feature) != true.
+  uint32_t m_notRoadCount;
+
+  Processor(generator::SrtmTileManager & manager)
+    : m_srtmManager(manager), m_roadCount(0), m_emptyRoadCount(0), m_roadPointCount(0), m_notRoadCount(0)
+  {
+  }
+
+  void operator()(FeatureType const & f, uint32_t const & id)
+  {
+    f.ParseBeforeStatistic();
+    if (!GetBicycleModel().IsRoad(f))
+    {
+      ++m_notRoadCount;
+      return;
+    }
+
+    f.ParseGeometry(FeatureType::BEST_GEOMETRY);
+    uint32_t const pointCount = f.GetPointsCount();
+    if (pointCount == 0)
+    {
+      ++m_emptyRoadCount;
+      return;
+    }
+
+    ++m_roadCount;
+
+    m_roadPointCount += pointCount;
+
+    for (int i = 0; i < pointCount; ++i)
+      m_uniqueRoadPoints.insert(RoughPoint(f.GetPoint(i)));
+
+    // Feature length and feature segment length.
+    double realFeatureLengthMeters = 0.0;
+    for (int i = 0; i < pointCount - 1; ++i)
+    {
+      // Feature segment length.
+      double const realSegmentLengthMeters = MercatorBounds::DistanceOnEarth(f.GetPoint(i), f.GetPoint(i + 1));
+      m_segLength[static_cast<uint32_t>(floor(realSegmentLengthMeters))]++;
+
+      // Feature length.
+      realFeatureLengthMeters += realSegmentLengthMeters;
+    }
+    m_featureLength[static_cast<uint32_t>(floor(realFeatureLengthMeters))]++;
```

Comment:
No need in floor call - cast to uint32_t automatically throws away fractional part.